### PR TITLE
LOG4J-2518 LOG4J2-2433 Coroutines Support

### PR DESCRIPTION
**Pull request by rocketraman**
Add support for coroutines by fixing [LOG4J-2518](https://issues.apache.org/jira/browse/LOG4J2-2518) and [LOG4J2-2433](https://issues.apache.org/jira/browse/LOG4J2-2433).

Suspend functions inside lambda functions are supported by making the lambdas inline, and ThreadContext support is added to integrate the thread-local ThreadContext with coroutines context.

@jvz Can you take a look please?

### CODE REVIEW

File: log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt [hunk 1]
```diff
@@ -117,19 +117,19 @@ class KotlinLogger(val delegate: ExtendedLogger) {
     delegate.logIfEnabled(FQCN, level, null, msg, t)
   }
 
-  fun log(level: Level, supplier: () -> Any?) {
+  inline fun log(level: Level, supplier: () -> Any?) {
```
Comment (jvz) [inline log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt]: Does this affect the caller information in any breaking way?

Comment (rocketraman) [reply log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt]: My understanding is that this is both source and binary compatible. I can't say that with 100% certainty but I'll ask on the Kotlin Slack.

Comment (rocketraman) [reply log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt]: Yes, confirmed that it is.

Comment (jvz) [reply log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt]: I mean in the caller location pattern. If you have a pattern layout with `%C`, for example, you can get the name of the caller class with line numbers and things like that. That feature relies upon knowing which stack frames can be safely skipped to find the caller of the logger. By inlining things, you're changing the stack, so you should verify that caller info/method info still work.

**Comment by jvz** [comment]
Looks like a test failure in `LoggerCoroutinesTest`?

**Comment by rocketraman** [comment]
Hmm, that's odd. Works for me here. Let me see what is different on the build machine.

**Comment by rocketraman** [comment]
@jvz One approach is to do the message construction in the Kotlin code before delegating to log4j2. For example:

```
  inline fun error(supplier: () -> Any?) {
    if(delegate.isEnabled(Level.ERROR)) {
      val message = delegate.getMessageFactory<MessageFactory>().newMessage(supplier.invoke())
      delegate.logMessage(FQCN, Level.ERROR, null, message, null)
    }
  }
```

With this approach the supplier is invoked before the delegate, so suspending works correctly. However, it does require copying some machinery -- the isEnabled check and the message factory call -- out of the delegate. Overall though, it doesn't seem like a bad solution. Thoughts?

File: log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt (hunk 1 above)
Comment (rocketraman) [reply log4j-api-kotlin/src/main/kotlin/org/apache/logging/log4j/kotlin/KotlinLogger.kt]: @jvz No, the caller location seems to work the same as before -- the trace extension still doesn't do the right thing, but that was already a known issue.

**Comment by jvz** [comment]
If you identify an awkward API in log4j-api, we can always work on improving that upstream and later using it in version 1.1.x.

Feel free to implement that way you're describing. I'm mostly concerned about preserving caller location info where possible since some people rely on that feature (despite it causing logging to slow down by quite a bit). Now if we could use some sort of compiler feature or macros to implement the caller info, then we can safely inline a lot of things while also supporting caller info without much of a performance penalty.

**Comment by rocketraman** [comment]
> Now if we could use some sort of compiler feature or macros to implement the caller info, then we can safely inline a lot of things while also supporting caller info without much of a performance penalty.

Yeah, I'm pretty sure the Kotlin compiler plugin capabilities that are upcoming will allow us to do this.

> I'm mostly concerned about preserving caller location info where possible since some people rely on that feature

As per my other comment, this seems to work fine with the inlining, with the exception of the runInTrace methods. However, those are broken because `ExtendedLogger` does not expose running `traceExit` and `catching` with the FQCN, and was already a know issue/limitation. I had already opened https://issues.apache.org/jira/browse/LOG4J2-2275 to track this.

> Feel free to implement that way you're describing.

Great, I'll work on that over the next couple of days.

**Comment by rocketraman** [comment]
> As per my other comment, this seems to work fine with the inlining

Nope, my mistake, caller information does break with inlining the supplier :-(

**Comment by rocketraman** [comment]
New pull https://github.com/apache/logging-log4j-kotlin/pull/8 submitted without the inlining, until we've decided how to handle this without breaking caller location.

**Comment by jvz** [comment]
Closing this in favor of the existing implementation.

**Pull request closed without merging**